### PR TITLE
feat(gateway): improve GO API interface, remove Writable API

### LINKED ISSUE #61: gateway: post-extraction GO API cleanup
We're currently working on extracting the Gateway code from `kubo` to `go-libipfs` (https://github.com/ipfs/kubo/issues/8524). After the extraction, we should improve the code, such that it is more useable. There are currently a few points that, in my opinion, make the Gateway handler harder to use than it should:

1. It requires more APIs than necessary.
	- From `NodeAPI.Block()`, we only use `Get` and `Stat`, but we are requiring the full interface (`Get`, `Put`, `Rm`, `Stat`).
	- From `NodeAPI.Dag()`, we only use `Get`, but we are requiring the full `APIDagService` interface, which contains a lot more functions.
2. The handler requires both an online and offline APIs. The offline API is only used once with the `Dag().Get` function. Perhaps we could add an option to that method `Offline: true` or `LocalOnly: true` or `NoFetch: true` instead. This would allow us to only require the consumers of the Gateway code to provide one Node API.

Things to do:

- [x] Define a minimal set of APIs required by the Gateway. https://github.com/ipfs/go-libipfs/issues/61
- [x] Remove Writable API. https://github.com/ipfs/go-libipfs/issues/61 & https://github.com/ipfs/kubo/pull/9616
	- The Writable API is undocumented and experimental feature of Kubo. We want to support a POST/PUT API in the future that will be incompatible. So it is a good idea to remove it right now and do not carry this bagage.

### PR DESCRIPTION
**Pull request by hacdias**
Closes #61

This PR does the following:

1. Simplifies `NodeAPI` (renamed `API`), such that we only require the bare minimum functions we need on the gateway. These have each the minimum required amount of functions required. This will allow other developers to use our gateway code much more easily.
2. Consolidates the `offlineAPI` as a single function on `API` (`IsCached`).
3. Removes the Writable Gateway. The code for the Writable Gateway is moved to Kubo codebase and deprecated https://github.com/ipfs/kubo/issues/9622. See https://github.com/ipfs/kubo/pull/9616.

Please **note**: gateway sharness tests are expected to fail since they are run against Kubo's `master` version. The PR in Kubo with this changes can be viewed at: https://github.com/ipfs/kubo/pull/9616

### CODE REVIEW

**Comment by hacdias** [comment]
**Note**: I would also prefer to completely remove the Offline API and instead adding an option to Dag().Get() to do LocalOnly.

**Comment by hacdias** [comment]
I converted this to draft because I will be doing to more things: removing the Writable gateway and updating the interface to make it more simple.

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]: nits:
- Since we have a separate `ResolvePath`, does it make sense to change this to require  [`Resolved`](https://github.com/ipfs/interface-go-ipfs-core/blob/master/path/path.go#L44) path instead?
- It returns a `BlockReader`, not a block bytes

```suggestion
	BlockReader(context.Context, path.Resolved) (io.Reader, error)
```

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]: Confusing name, `GetUnixFs` returns `Node` too. Make this explicit:

```suggestion
	GetIPLDNode(context.Context, cid.Cid) (ipld.Node, error)
```

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]: Gateway ignores `BlockStat` – we only care if it errors (block missing from local blockstore) or not, so it works as a simple boolean. Make it explicit and clear:

```suggestion
	IsCached(context.Context, path.Path) (bool)
```

File: gateway/handler_car.go
Comment (lidel) [inline gateway/handler_car.go]: :thought_balloon: Interesting, we don't need  `ipld.Node`  returned by `Get[IPLD]Node` here, [`readStore` only uses `blocks.Block`](https://github.com/ipld/go-car/blob/268b65bab2b7f45aa05494b943992f5b9bbab75e/car.go#L26).


:point_right:   Are we able to simplify and switch to `blocks.Block` created with `BlockReader`,
removing dependency on IPLD if someone wants to only use this library for hosting CARs and Blocks?



(:thought_balloon:  This is short term cleanup. Long term, CAR response handler will be refactored  again, and in the case of bifrost-gateway most likely we will be doing pass-through of a CAR generated  on Saturn nodes.)

File: gateway/handler_codec.go
Comment (lidel) [inline gateway/handler_codec.go]: :thought_balloon:  This seems to be the only place we use `Get[IPLD]Node`.

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]: We only use this for IPNS, make it explicit, and require a valid Cid instead of a string:

```suggestion
	// GetIPNSRecord retrieves the best IPNS record for a given CID (libp2p-key) from the routing system.
	GetIPNSRecord(context.Context, cid.Cid) ([]byte, error)
```

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]:
```suggestion
	GetUnixFsNode(context.Context, path.Path) (files.Node, error)
```

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]:
```suggestion
	LsUnixFsDir(context.Context, path.Path, ...options.UnixfsLsOption) (<-chan iface.DirEntry, error)
```

File: gateway/handler.go
Comment (lidel) [inline gateway/handler.go:320]: :information_source:  I've filled https://github.com/ipfs/kubo/issues/9622 to decide if we move this code back to kubo and do soft deprecation, or break existing users.

**Review by lidel — CHANGES_REQUESTED** [review]
Thank you for identifying these @hacdias. 

I've dropped some ideas to simplify it further + avoid confusing names + follow up for cleaning up writable mode in Kubo.

File: gateway/handler_codec.go
Comment (hacdias) [reply gateway/handler_codec.go]: Yeah, I don't think we can get "rid of" IPLD since we have the conversions. Perhaps we can do something along the lines of: GetBlock, convert it to a node ourselves, etc. Let me try.

Comment (hacdias) [reply gateway/handler_codec.go]: Actually managed to remove it :D

File: gateway/handler_unixfs_dir.go
Comment (hacdias) [inline gateway/handler_unixfs_dir.go:114]: Moved this optimization to Kubo's codebase. Each gateway implementation should use whatever they want, some might be faster, some slower. It's the implementers choice. This also allowed me to simplify the API signature to remove the options. 

```diff
- LsUnixFsDir(context.Context, path.Path, ...options.UnixfsLsOption) (<-chan iface.DirEntry, error)
+ LsUnixFsDir(context.Context, path.Resolved) (<-chan iface.DirEntry, error)
```

File: gateway/handler_unixfs_dir.go
Comment (hacdias) [inline gateway/handler_unixfs_dir.go:71]: I updated the signature of the getter for UnixFS files:

```diff
- GetUnixFsNode(context.Context, path.Path) (files.Node, error)
+ GetUnixFsNode(context.Context, path.Resolved) (files.Node, error)
```

We already have `ResolvePath` at our disposal. Let's simplify. The change in this code is to reflect that.

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go:21]: nit:  something like `IPFSBackend` would be more descriptive, but can be changed later, won't block on this one thing  (this interface will most likely get revamp in near future anyway)

**Review by lidel — APPROVED** [review]
Thank you @hacdias, moving `Gateway.Writable` to https://github.com/ipfs/kubo/pull/9616 is :+1:  
Merging to unblock other PRs based on this.

**Pull request merged**
